File: old/address.cpp
/* SPDX-License-Identifier: MPL-2.0 */

#include "precompiled.hpp"
#include "macros.hpp"
#include "address.hpp"
#include "ctx.hpp"
#include "err.hpp"
#include "tcp_address.hpp"
#include "udp_address.hpp"
#include "ipc_address.hpp"
#include "tipc_address.hpp"
#include "ws_address.hpp"

#if defined ZMQ_HAVE_VMCI
#include "vmci_address.hpp"
#endif

#include <string>
#include <sstream>

zmq::address_t::address_t (const std::string &protocol_,
                           const std::string &address_,
                           ctx_t *parent_) :
    protocol (protocol_), address (address_), parent (parent_)
{
    resolved.dummy = NULL;
}

zmq::address_t::~address_t ()
{
    if (protocol == protocol_name::tcp) {
        LIBZMQ_DELETE (resolved.tcp_addr);
    } else if (protocol == protocol_name::udp) {
        LIBZMQ_DELETE (resolved.udp_addr);
    }
#ifdef ZMQ_HAVE_WS
    else if (protocol == protocol_name::ws) {
        LIBZMQ_DELETE (resolved.ws_addr);
    }
#endif

#ifdef ZMQ_HAVE_WSS
    else if (protocol == protocol_name::wss) {
        LIBZMQ_DELETE (resolved.ws_addr);
    }
#endif

#if defined ZMQ_HAVE_IPC
    else if (protocol == protocol_name::ipc) {
        LIBZMQ_DELETE (resolved.ipc_addr);
    }
#endif
#if defined ZMQ_HAVE_TIPC
    else if (protocol == protocol_name::tipc) {
        LIBZMQ_DELETE (resolved.tipc_addr);
    }
#endif
#if defined ZMQ_HAVE_VMCI
    else if (protocol == protocol_name::vmci) {
        LIBZMQ_DELETE (resolved.vmci_addr);
    }
#endif
}

int zmq::address_t::to_string (std::string &addr_) const
{
    if (protocol == protocol_name::tcp && resolved.tcp_addr)
        return resolved.tcp_addr->to_string (addr_);
    if (protocol == protocol_name::udp && resolved.udp_addr)
        return resolved.udp_addr->to_string (addr_);
#ifdef ZMQ_HAVE_WS
    if (protocol == protocol_name::ws && resolved.ws_addr)
        return resolved.ws_addr->to_string (addr_);
#endif
#ifdef ZMQ_HAVE_WSS
    if (protocol == protocol_name::wss && resolved.ws_addr)
        return resolved.ws_addr->to_string (addr_);
#endif
#if defined ZMQ_HAVE_IPC
    if (protocol == protocol_name::ipc && resolved.ipc_addr)
        return resolved.ipc_addr->to_string (addr_);
#endif
#if defined ZMQ_HAVE_TIPC
    if (protocol == protocol_name::tipc && resolved.tipc_addr)
        return resolved.tipc_addr->to_string (addr_);
#endif
#if defined ZMQ_HAVE_VMCI
    if (protocol == protocol_name::vmci && resolved.vmci_addr)
        return resolved.vmci_addr->to_string (addr_);
#endif

    if (!protocol.empty () && !address.empty ()) {
        std::stringstream s;
        s << protocol << "://" << address;
        addr_ = s.str ();
        return 0;
    }
    addr_.clear ();
    return -1;
}

zmq::zmq_socklen_t zmq::get_socket_address (fd_t fd_,
                                            socket_end_t socket_end_,
                                            sockaddr_storage *ss_)
{
    zmq_socklen_t sl = static_cast<zmq_socklen_t> (sizeof (*ss_));

    const int rc =
      socket_end_ == socket_end_local
        ? getsockname (fd_, reinterpret_cast<struct sockaddr *> (ss_), &sl)
        : getpeername (fd_, reinterpret_cast<struct sockaddr *> (ss_), &sl);

    return rc != 0 ? 0 : sl;
}

use std::ptr::null_mut;
use std::ffi::CString;

mod precompiled;
mod macros;
mod address;
mod ctx;
mod err;
mod tcp_address;
mod udp_address;
mod ipc_address;
mod tipc_address;
mod ws_address;

#[cfg(feature = "vmci")]
mod vmci_address;

struct Address {
    protocol: String,
    address: String,
    parent: *mut ctx::Ctx,
    resolved: ResolvedAddress,
}

enum ResolvedAddress {
    Dummy(*mut ()),
    Tcp(*mut tcp_address::TcpAddress),
    Udp(*mut udp_address::UdpAddress),
    // Add other address types as needed
}

impl Address {
    fn new(protocol: &str, address: &str, parent: *mut ctx::Ctx) -> Address {
        Address {
            protocol: protocol.to_string(),
            address: address.to_string(),
            parent,
            resolved: ResolvedAddress::Dummy(null_mut()),
        }
    }
}

impl Drop for Address {
    fn drop(&mut self) {
        match self.protocol.as_str() {
            "tcp" => {
                if let ResolvedAddress::Tcp(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            "udp" => {
                if let ResolvedAddress::Udp(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            _ => {}
        }
    }
}

File: src/address.rs
use crate::context::Context;
use crate::err::ZmqError;
#[cfg(all(feature = "ipc", unix))]
use crate::ipc_address;
use crate::sockaddr_storage::{zmq_socklen_to_c_int, ZmqSockaddrStorage, ZmqSocklen};
use crate::socket::{ZmqPlatformSocket, ZmqSockAddrIn, ZmqSockAddrIn6};
use crate::tcp_address::TcpAddress;
#[cfg(unix)]
use crate::tipc_address;
use crate::types::ZmqSaFamily;
use crate::udp_address::UdpAddress;
#[cfg(feature = "vmci")]
use crate::vmci_address;
#[cfg(feature = "ws")]
use crate::ws_address;
use std::os::raw::c_void;
use std::ptr::null_mut;

pub struct Address<'a> {
    protocol: String,
    address: String,
    parent: &'a mut Context,
    resolved: ResolvedAddress<'a>,
}

enum ResolvedAddress<'a> {
    Dummy(*mut c_void),
    Tcp(&'a mut TcpAddress),
    Udp(&'a mut UdpAddress),
    #[cfg(feature = "ws")]
    Ws(*mut ws_address::WsAddress),
    #[cfg(feature = "wss")]
    Wss(*mut ws_address::WsAddress),
    #[cfg(all(feature = "ipc", unix))]
    Ipc(*mut ipc_address::IpcAddress),
    #[cfg(all(feature = "tipc", target_os = "linux"))]
    Tipc(*mut tipc_address::TipcAddress),
    #[cfg(feature = "vmci")]
    Vmci(*mut vmci_address::VmciAddress),
}

impl<'a> Address<'a> {
    pub fn new(protocol: &str, address: &str, parent: &mut Context) -> Address<'a> {
        Address {
            protocol: protocol.to_string(),
            address: address.to_string(),
            parent,
            resolved: ResolvedAddress::Dummy(null_mut()),
        }
    }

    pub fn to_string(&self) -> Result<String, ZmqError> {
        match self.protocol.as_str() {
            "tcp" => {
                if let ResolvedAddress::Tcp(addr) = self.resolved {
                    addr.to_string()
                } else {
                    Err(ZmqError::ParsingError("Invalid TCP address".to_string()))
                }
            }
            "udp" => {
                if let ResolvedAddress::Udp(addr) = self.resolved {
                    addr.to_string()
                } else {
                    Err(ZmqError::ParsingError("Invalid UDP address".to_string()))
                }
            }
            #[cfg(feature = "ws")]
            "ws" => {
                if let ResolvedAddress::Ws(addr) = self.resolved {
                    unsafe { (*addr).to_string() }
                } else {
                    Err(())
                }
            }
            #[cfg(feature = "wss")]
            "wss" => {
                if let ResolvedAddress::Wss(addr) = self.resolved {
                    unsafe { (*addr).to_string() }
                } else {
                    Err(())
                }
            }
            #[cfg(feature = "ipc")]
            "ipc" => {
                if let ResolvedAddress::Ipc(addr) = self.resolved {
                    unsafe { (*addr).to_string() }
                } else {
                    Err(())
                }
            }
            #[cfg(any(feature = "linux", feature = "vxworks"))]
            "tipc" => {
                if let ResolvedAddress::Tipc(addr) = self.resolved {
                    unsafe { (*addr).to_string() }
                } else {
                    Err(())
                }
            }
            #[cfg(feature = "vmci")]
            "vmci" => {
                if let ResolvedAddress::Vmci(addr) = self.resolved {
                    unsafe { (*addr).to_string() }
                } else {
                    Err(())
                }
            }
            _ => {
                if !self.protocol.is_empty() && !self.address.is_empty() {
                    Ok(format!("{}://{}", self.protocol, self.address))
                } else {
                    Err(())
                }
            }
        }
    }
}

impl Drop for Address<'_> {
    fn drop(&mut self) {
        match self.protocol.as_str() {
            "tcp" => {
                if let ResolvedAddress::Tcp(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            "udp" => {
                if let ResolvedAddress::Udp(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            #[cfg(feature = "ws")]
            "ws" => {
                if let ResolvedAddress::Ws(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            #[cfg(feature = "wss")]
            "wss" => {
                if let ResolvedAddress::Wss(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            #[cfg(feature = "ipc")]
            "ipc" => {
                if let ResolvedAddress::Ipc(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            #[cfg(any(feature = "linux", feature = "vxworks"))]
            "tipc" => {
                if let ResolvedAddress::Tipc(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            #[cfg(feature = "vmci")]
            "vmci" => {
                if let ResolvedAddress::Vmci(addr) = self.resolved {
                    unsafe { Box::from_raw(addr) };
                }
            }
            _ => {}
        }
    }
}

pub fn get_socket_address(
    fd: ZmqPlatformSocket,
    socket_end: SocketEnd,
    ss: &mut ZmqSockaddrStorage,
) -> usize {
    let mut sock_len = size_of::<ZmqSockaddrStorage>() as ZmqSocklen;
    let mut sock_len_c_int = zmq_socklen_to_c_int(sock_len);

    let rc = match socket_end {
        SocketEnd::Local => unsafe {
            libc::getsockname(fd, ss as *mut _ as *mut _, &mut sock_len_c_int)
        },
        SocketEnd::Remote => unsafe {
            libc::getpeername(fd, ss as *mut _ as *mut _, &mut sock_len_c_int)
        },
    };

    if rc != 0 {
        0
    } else {
        sock_len as usize
    }
}

pub enum SocketEnd {
    Local,
    Remote,
}

pub fn get_socket_name(fd: ZmqPlatformSocket, socket_end: SocketEnd) -> Result<String, ZmqError> {
    // TODO: i believe this part is getting the tcp address in the form tcp://<addr>:port?
    let mut ss: ZmqSockaddrStorage = ZmqSockaddrStorage::new();
    let socklen = get_socket_address(fd, socket_end, &mut ss);
    if socklen == 0 {
        return Err(ZmqError::ParsingError(
            "Failed to get socket address".to_string(),
        ));
    }

    match ss.ss_family {
        ZmqSaFamily::AF_INET => {
            let sockaddr = ss.__ss_pad2.as_mut_ptr() as *mut ZmqSockAddrIn;
            let tcp_addr = TcpAddress::from_sockaddr(sockaddr);

            let sa = unsafe { *(ss.__ss_pad2.as_ptr() as *const ZmqSockAddrIn) };
            let addr = TcpAddress::from_sockaddr(&sa);
            addr.to_string()
        }
        ZmqSaFamily::AF_INET6 => {
            let sai = unsafe { *(ss.__ss_pad2.as_ptr() as *const ZmqSockAddrIn6) };
            let tcp_addr = TcpAddress::from_sockaddr_in6(&sai);
            tcp_addr.to_string()
        }
        _ => Err(ZmqError::ParsingError(
            "Unsupported address family".to_string(),
        )),
    }
}
